server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java

```diff
@@ -0,0 +1,349 @@
+                this.lastCommittedSegmentInfos = Lucene.readSegmentInfos(directory);
+                this.translogStats = translogStats;
+                this.seqNoStats = seqNoStats == null ? buildSeqNoStats(lastCommittedSegmentInfos) : seqNoStats;
+                reader = wrapReader(DirectoryReader.open(directory), config.getShardId());
+                this.indexCommit = reader.getIndexCommit();
+                this.searcherManager = new SearcherManager(reader, new SearcherFactory());
+                this.indexWriterLock = indexWriterLock;
+                success = true;
+            } finally {
+                if (success == false) {
+                    IOUtils.close(reader, indexWriterLock, store::decRef);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e); // this is stupid
+        }
+    }
+
+    @Override
+    protected void closeNoLock(String reason, CountDownLatch closedLatch) {
+        if (isClosed.compareAndSet(false, true)) {
+            try {
+                IOUtils.close(searcherManager, indexWriterLock, store::decRef);
+            } catch (Exception ex) {
+                logger.warn("failed to close searcher", ex);
+            } finally {
+                closedLatch.countDown();
+            }
+        }
+    }
+
+    protected DirectoryReader wrapReader(DirectoryReader reader, ShardId shardId) throws IOException {
+        return ElasticsearchDirectoryReader.wrap(reader, shardId);
+    }
+
+    public static SeqNoStats buildSeqNoStats(SegmentInfos infos) {
+        final SequenceNumbers.CommitInfo seqNoStats =
+            SequenceNumbers.loadSeqNoInfoFromLuceneCommit(infos.userData.entrySet());
+        long maxSeqNo = seqNoStats.maxSeqNo;
+        long localCheckpoint = seqNoStats.localCheckpoint;
+        return new SeqNoStats(maxSeqNo, localCheckpoint, localCheckpoint);
+    }
+
+    @Override
+    public GetResult get(Get get, BiFunction<String, SearcherScope, Searcher> searcherFactory) throws EngineException {
+        return getFromSearcher(get, searcherFactory, SearcherScope.EXTERNAL);
+    }
+
+    @Override
+    protected ReferenceManager<IndexSearcher> getReferenceManager(SearcherScope scope) {
+        return searcherManager;
+    }
+
+    @Override
+    protected SegmentInfos getLastCommittedSegmentInfos() {
+        return lastCommittedSegmentInfos;
+    }
+
+    @Override
+    public String getHistoryUUID() {
+        return lastCommittedSegmentInfos.userData.get(Engine.HISTORY_UUID_KEY);
+    }
+
+    @Override
+    public long getWritingBytes() {
+        return 0;
+    }
+
+    @Override
+    public long getIndexThrottleTimeInMillis() {
+        return 0;
+    }
+
+    @Override
+    public boolean isThrottled() {
+        return false;
+    }
+
+    @Override
+    public IndexResult index(Index index) {
+        throw new UnsupportedOperationException();
```

Comment:
we should assert this is never called (same for the other places here where `UnsupportedOperationException` is thrown), as this indicates a bug.